index.js

```diff
@@ -8,42 +8,50 @@ class MongoToSqs {
     this.queueUrl = options.queueUrl; // https://sqs.us-east-1.amazonaws.com/000000000000/my-queue
     this.formatPayload = options.formatPayload || (doc => doc);
     this.concurrency = options.concurrency || 2500;
+    this.started = false;
+    this.cursorEnd = false;
     // TODO: this.sqsBatchSize - instead of sending every sqs doc individually
   }
 
   sendToQueue(payload) {
     return this.sqs.sendMessage({
       MessageBody: JSON.stringify(payload),
-      DelaySeconds: 0,
       QueueUrl: this.queueUrl
     }).promise();
   }
 
   start() {
+    if (this.started) {
+      throw new Error('SQS loader has already started.');
+    }
+    this.started = true;
     let count = 0;
-    return new Promise((resolve, reject) => {
+    return new Promise(resolve => {
+      const handleSQSResponse = (done) => {
+        count -= 1;
+        done();
+        // if all docs have successfully queued onto SQS
+        if (this.cursorEnd && count === 0) {
+          resolve();
+        }
+      };
       this.cursor.pipe(through2Concurrent.obj({
         maxConcurrency: this.concurrency
       }, (doc, encoding, done) => {
         count += 1;
         const payload = this.formatPayload(doc);
         this.sendToQueue(payload)
           .then(() => {
-            count -= 1;
-            done();
+            handleSQSResponse(done);
           })
           .catch(err => {
-            count -= 1;
             console.log('sendToQueue:', doc, err);
-            done();
+            handleSQSResponse(done);
           });
       }));
       // done reading from mongo collection
       this.cursor.on('end', () => {
-        // wait until all docs have been processed
-        setInterval(() => {
-          if (count === 0) resolve();
-        }, 1000);
+        this.cursorEnd = true;
       });
```

Comment:
Don't we need to handle error here or is it intentionally left out?
```js
this.cursor.on('error', err => {
  reject(err);
});
```

Reply:
Since the cursor is passed in to this module, I'm thinking handling cursor error is the responsibility of whoever created the cursor.

Reply:
Actually n/m I think you're right.

Reply:
added